worker/changestream/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package eventqueue
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/changestream"
+)
+
+// Logger represents the logging methods called.
+type Logger interface {
+	Infof(message string, args ...interface{})
+	Tracef(message string, args ...interface{})
+	IsTraceEnabled() bool
+}
+
+// Stream represents a way to get change events.
+type Stream interface {
+	// Changes returns a channel for a given namespace (database).
+	Changes() <-chan changestream.ChangeEvent
+}
+
+type subscription struct {
+	id      int
+	changes chan changestream.ChangeEvent
+	topics  map[string]struct{}
+
+	unsubscriber func()
+}
+
+// Unsubscribe removes the subscription from the event queue.
+func (s *subscription) Unsubscribe() {
+	s.unsubscriber()
+}
+
+// Changes returns a channel that will receive events from the event queue
+// for the subscription.
+func (s *subscription) Changes() <-chan changestream.ChangeEvent {
+	return s.changes
+}
+
+type eventFilter struct {
+	subscriptionID int
+	changeMask     changestream.ChangeType
+	filter         func(changestream.ChangeEvent) bool
+}
+
+// EventQueue is a queue of events that can be subscribed to.
+type EventQueue struct {
+	tomb   tomb.Tomb
+	stream Stream
+	logger Logger
+
+	mutex             sync.Mutex
+	subscriptions     map[int]*subscription
+	subscriptionsByNS map[string][]*eventFilter
+
+	actions chan func()
+}
+
+// New creates a new EventQueue that will use the Stream for events.
+func New(stream Stream, logger Logger) *EventQueue {
+	queue := &EventQueue{
+		stream:            stream,
+		logger:            logger,
+		subscriptions:     make(map[int]*subscription),
+		subscriptionsByNS: make(map[string][]*eventFilter),
+		actions:           make(chan func()),
+	}
+
+	queue.tomb.Go(queue.loop)
+
+	return queue
+}
+
+// Subscribe creates a new subscription to the event queue.
+func (s *EventQueue) Subscribe(opts ...changestream.SubscriptionOption) (changestream.Subscription, error) {
+	if len(opts) == 0 {
+		return nil, errors.Errorf("no subscription options specified")
+	}
+
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	// Create a new subscription and assign a unique ID to it.
+	subID := len(s.subscriptions)
```

Comment:
This shouldn't be length, as there is a potential to overwrite the subID after an unsubscription.